File: core/src/main/java/pcb/uwu/amount/derived/scalar/SolidAngle.java
package pcb.uwu.amount.derived.scalar;

import org.jetbrains.annotations.NotNull;
import pcb.uwu.core.BigDecimalAmount;
import pcb.uwu.core.CompositeUnitAmount;
import pcb.uwu.core.Magnitude;
import pcb.uwu.core.UnitAmount;
import pcb.uwu.unit.derived.scalar.SolidAngleUnit;

import java.math.BigDecimal;
import java.math.MathContext;

import static pcb.uwu.utils.UnitAmountUtils.dividedByScalar;
import static pcb.uwu.utils.UnitAmountUtils.getAmountIn;
import static pcb.uwu.utils.UnitAmountUtils.minusAmount;
import static pcb.uwu.utils.UnitAmountUtils.multipliedByScalar;
import static pcb.uwu.utils.UnitAmountUtils.plusAmount;

public class SolidAngle extends CompositeUnitAmount<SolidAngleUnit> {

	// region constructors

	public SolidAngle(Number value, SolidAngleUnit unit) {
		super(value, unit);
	}

	public SolidAngle(Number value, Magnitude magnitude, SolidAngleUnit unit) {
		super(value, magnitude, unit);
	}

	public SolidAngle(String value, SolidAngleUnit unit) {
		super(value, unit);
	}

	public SolidAngle(String value, Magnitude magnitude, SolidAngleUnit unit) {
		super(value, magnitude, unit);
	}

	public SolidAngle(BigDecimal value, SolidAngleUnit unit) {
		super(value, unit);
	}

	public SolidAngle(BigDecimal value, Magnitude magnitude, SolidAngleUnit unit) {
		super(value, magnitude, unit);
	}

	public SolidAngle(BigDecimalAmount amount, SolidAngleUnit unit) {
		super(amount, unit);
	}

	public SolidAngle(BigDecimalAmount amount, Magnitude magnitude, SolidAngleUnit unit) {
		super(amount, magnitude, unit);
	}


	// endregion

	// region implement UnitAmount

	@NotNull
	@Override
	public SolidAngle plus(@NotNull UnitAmount<SolidAngleUnit> other) {
		return new SolidAngle(plusAmount(this, other), getUnit());
	}

	@NotNull
	@Override
	public SolidAngle minus(@NotNull UnitAmount<SolidAngleUnit> other) {
		return new SolidAngle(minusAmount(this, other), getUnit());
	}

	@Override
	public SolidAngle multiply(BigDecimal other, MathContext mathContext) {
		return new SolidAngle(multipliedByScalar(this, other, mathContext), getUnit());
	}

	@Override
	public SolidAngle div(BigDecimal other, MathContext mathContext) {
		return new SolidAngle(dividedByScalar(this, other, mathContext), getUnit());
	}

	@Override
	public SolidAngle into(SolidAngleUnit unit) {
		return new SolidAngle(getAmountIn(this, unit), unit);
	}

	// endregion

	// region composition

	
	
	// endregion
}

File: core/src/main/java/pcb/uwu/amount/derived/scalar/SolidAngle.kt
package pcb.uwu.amount.derived.scalar

import pcb.uwu.amount.base.LuminousIntensity
import pcb.uwu.amount.derived.optics.LuminousFlux
import pcb.uwu.core.CompositeUnitAmount
import pcb.uwu.core.Magnitude
import pcb.uwu.core.Magnitude.NATURAL
import pcb.uwu.core.UnitAmount
import pcb.uwu.unit.derived.optics.LuminousFluxUnit
import pcb.uwu.unit.scalar.SolidAngleUnit
import pcb.uwu.utils.UnitAmountUtils

open class SolidAngle : CompositeUnitAmount<SolidAngleUnit>
{
    @JvmOverloads
    constructor(amount: Number,
                magnitude: Magnitude = NATURAL,
                unit: SolidAngleUnit)
            : super(amount, magnitude, unit)

    @JvmOverloads
    constructor(amount: String,
                magnitude: Magnitude = NATURAL,
                unit: SolidAngleUnit)
            : super(amount, magnitude, unit)

    // region UnitAmount

    @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
    override fun plus(solidAngle: UnitAmount<SolidAngleUnit>) =
        SolidAngle(amount = this.amount + (solidAngle to this.unit).amount,
                   unit = this.unit)

    @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
    override fun minus(solidAngle: UnitAmount<SolidAngleUnit>) =
        SolidAngle(amount = this.amount - (solidAngle to this.unit).amount,
                   unit = this.unit)

    override fun times(scalar: Number) =
        SolidAngle(amount = this.amount * scalar,
                   unit = this.unit)

    override fun div(scalar: Number) =
        SolidAngle(amount = this.amount / scalar,
                   unit = this.unit)

    override fun to(unit: SolidAngleUnit) =
        SolidAngle(amount = UnitAmountUtils.getAmountIn(unitAmount = this, newUnit = unit),
                   unit = unit)

    // endregion

    // region composition

    open operator fun times(luminousIntensity: LuminousIntensity) =
        LuminousFlux(amount = this.amount * luminousIntensity.amount,
                     unit = LuminousFluxUnit(solidAngleUnit = this.unit,
                                             luminousIntensityUnit = luminousIntensity.unit))

    // endregion
}